docs/src/proposals/program-instruction-macro.md

```diff
@@ -0,0 +1,224 @@
+rpc methods could return decoded instruction details, these custom solutions
+would be unnecessary.
+
+We can deserialize instruction data using a program's Instruction enum, but
+decoding the account-key list into human-readable identifiers requires manual
+parsing. Our current Instruction enums have that account information, but only
+in variant docs.
+
+Similarly, we have instruction constructor functions that duplicate nearly all
+the information in the enum, but we can't generate that constructor from the
+enum definition because the list of account references is in code comments.
+
+Also, Instruction docs can vary between implementations, as there is no
+mechanism to ensure consistency.
+
+## Proposed Solution
+
+Move the data from code comments to attributes, such that the constructors
+can be generated, and include all the documentation from the enum definition.
+
+Here is an example of an Instruction enum using the new accounts format:
+
+```rust,ignore
+#[instructions(test_program::id())]
+pub enum TestInstruction {
+    /// Transfer lamports
+    #[accounts(
+        from_account(signer, writable, desc = "Funding account"),
+        to_account(writable, desc = "Recipient account"),
+    )]
+    Transfer {
+        lamports: u64,
+    },
+
+    /// Provide M of N required signatures
+    #[accounts(
+        data_account(writable, desc = "Data account"),
+        signer(signer, multiple, desc = "Signer"),
+    )]
+    Multisig,
+
+    /// Consumes a stored nonce, replacing it with a successor
+    #[accounts(
+        nonce_account(signer, writable, desc = "Nonce account"),
+        recent_blockhashes_sysvar(signer, writable, desc = "RecentBlockhashes sysvar"),
+        nonce_authority(signer, optional, desc = "Nonce authority"),
+    )]
+    AdvanceNonceAccount,
+}
+```
+
+An example of the generated TestInstruction with docs:
+```rust,ignore
+    pub enum TestInstruction {
+      /// Transfer lamports
+      ///
+      /// * Accounts expected by this instruction:
+      ///   0. `[writable, signer]` Funding account
+      ///   1. `[writable]` Recipient account
+      Transfer {
+          lamports: u64,
+      },
+
+      /// Provide M of N required signatures
+      ///
+      /// * Accounts expected by this instruction:
+      ///   0. `[writable]` Data account
+      ///   * (Multiple) `[signer]` Signers
+      Multisig,
+
+      /// Consumes a stored nonce, replacing it with a successor
+      ///
+      /// * Accounts expected by this instruction:
+      ///   0. `[writable, signer]` Nonce account
+      ///   1. [] RecentBlockhashes sysvar
+      ///   2. (Optional) `[signer]` Nonce authority
+      AdvanceNonceAccount,
+  }
+```
+
+Generated constructors:
+```rust,ignore
+/// Transfer lamports
+///
+/// * `from_account` - `[writable, signer]` Funding account
+/// * `to_account` - `[writable]` Recipient account
+pub fn transfer(from_account: &Pubkey, to_account: &Pubkey, lamports: u64) -> Instruction {
```

Comment:
I think `Pubkey` will play out better than `&Pubkey`.  It'll be a bunch of changes to the existing code, but I don't see a problem with that.

Reply:
All right. I'd like to implement that change separately from the macro application. Does that work?